cmd/upload.go

```diff
@@ -86,3 +118,174 @@ func uploadModule(cmd *cobra.Command, args []string) error {
+	}
+
+	validateCtx, cancelValidateCtx := context.WithTimeout(ctx, 15*time.Second)
+	defer cancelValidateCtx()
+	signingKeys, err := storageBackend.SigningKeys(validateCtx, flagProviderNamespace)
+	if err != nil {
+		return err
+	}
+
+	sumsBytes, err := os.ReadFile(flagFileSha256Sums)
+	if err != nil {
+		return fmt.Errorf("failed to read file at path %s: %w", flagFileSha256Sums, err)
+	}
+
+	// We expect the signature to be suffixed with the .sig extension
+	// https://developer.hashicorp.com/terraform/registry/providers/publishing#manually-preparing-a-release
+	p := fmt.Sprintf("%s.sig", flagFileSha256Sums)
+	sumsSigBytes, err := os.ReadFile(p)
+	if err != nil {
+		return fmt.Errorf("failed to read file at path %s: %w", p, err)
+	}
+
+	if err := signingKeys.IsValidSha256Sums(sumsBytes, sumsSigBytes); err != nil {
+		return err
+	}
+
+	// Upload provider binary .zip archives
+	if len(flagProviderArchivePaths) > 0 {
+		for _, archivePath := range flagProviderArchivePaths {
+			if err := uploadProviderReleaseFile(ctx, storageBackend, archivePath, flagProviderNamespace, sums.Name()); err != nil {
+				return err
+			}
+			_ = level.Info(logger).Log(
+				"msg", "successfully published provider binary",
+				"name", filepath.Base(archivePath),
+			)
+		}
+	} else {
+		baseDir := filepath.Dir(flagFileSha256Sums)
+		for _, entry := range sums.Entries {
+			archivePath := filepath.Join(baseDir, entry.FileName)
+			if err := uploadProviderReleaseFile(ctx, storageBackend, archivePath, flagProviderNamespace, sums.Name()); err != nil {
+				return err
+			}
+			_ = level.Info(logger).Log(
+				"msg", "successfully published provider binary",
+				"name", entry.FileName,
+			)
+		}
+	}
+
+	// Upload *_SHA256SUMS file
+	if err = uploadProviderReleaseFile(ctx, storageBackend, flagFileSha256Sums, flagProviderNamespace, sums.Name()); err != nil {
+		return err
+	}
+	_ = level.Info(logger).Log(
+		"msg", "successfully published provider SHA256SUMS file",
+		"name", filepath.Base(flagFileSha256Sums),
+	)
+
+	// Upload *_SHA256SUMS.sig file
+	signaturePath := fmt.Sprintf("%s.sig", flagFileSha256Sums)
+	if err = uploadProviderReleaseFile(ctx, storageBackend, signaturePath, flagProviderNamespace, sums.Name()); err != nil {
+		return err
+	}
+	_ = level.Info(logger).Log(
+		"msg", "successfully published provider SHA256SUMS.sig file",
+		"name", filepath.Base(signaturePath),
+	)
+
+	return nil
+}
+
+func validateShaSums(sums *core.Sha256Sums) error {
+	// Check whether the user has given archive paths to upload on the command line as flags.
+	// If not, we try to determine the locations of the provider zip archives based on the path of the *_SHA256SUMS file and the filenames in that file
+	if len(flagProviderArchivePaths) != 0 {
+		if len(sums.Entries) != len(flagProviderArchivePaths) {
+			return fmt.Errorf("the number of provided archive paths doesn't match the number of entries in %s", flagFileSha256Sums)
+		}
+
+	OUTER:
```

Comment:
Can we change this part to avoid using labeling?

Reply:
I don't know why I chose the complicated route when I could've used `break` instead. Fixed this in 5c55979